dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ProgramRuleEngine.java

```diff
@@ -220,6 +218,15 @@
 
     private List<ProgramRule> getImplementableRules( Program program )
     {
-        return programRuleService.getImplementableProgramRules( program, IMPLEMENTABLE_TYPES );
+        List<ProgramRule> permittedRules = new ArrayList<>();
+
+        permittedRules = programRuleService.getImplementableProgramRules( program, ProgramRuleActionType.getNotificationLinkedTypes() );
+
+        if ( permittedRules.isEmpty() )
+        {
+            return permittedRules;
+        }
```

Comment:
Lines 221 - 228 can be rewritten as: 
`
if ( programRuleService.getImplementableProgramRules( program, ProgramRuleActionType.getNotificationLinkedTypes().isEmpty() ) 
{
     return Collections.emptySet();
}
`
But it is only a minor thing.

However, I am not sure I follow the logic. So first you check whether there are some rules of NotificationLinkedTypes (so `sendmessage` and `schedulemessage`) and if yes, then you fetch the rules for implemented actions (so `sendmessage`, `schedulemessage` and `assign`)? Does it make sense? If yes, then fine, looks OK.

Another question is:
- How often is this method called?
- How often are the rules changed?

Based on the answers: Does it make sense to cache it?

Reply:
the logic is
we only need to trigger backend rule engine if we have ProgramRuleActionType.SENDMESSAGE/SCHEDLEMESSAGE.
So first it checks if these two types of rule actions exist or not. In case they exist then we need to take ProgramRuleActionType.ASSIGN as well in order to evaluate expressions correctly.
Secondly, it will be called at stage/enrollment completion. Rules do not change frequently but having them in the cache would increase the load on cache as there might be cases where there are more than 500 rules

Reply:
Hmm, if it can be up to 500 rules, then fetching it every time can be heavy. I think that completion is happening quite often, or? But OK, it is fine for me to leave it as it is and if there will be some performance issues in the future, then we can improve it. Maybe wait what @stian-sandvold has to say.

Reply:
I guess every rule in this case is usually quite unique @zubaira ? Not sure if there are any good ways to optimize this as of now? If we need to fetch 500 rules, we have to fetch 500 rules. Is there a performance issue known related to this?